### PR TITLE
Add verify_with_flags to let caller select script interpreter flags

### PR DESCRIPTION
**Pull request by ariard**
Enable checking of input scripts against standard flags at the transaction level.

See https://github.com/rust-bitcoin/rust-bitcoinconsensus/pull/15

### CODE REVIEW

**Comment by elichai** [comment]
FYI 
Can bitcoin's deserialization code dereference any of the pointers if the length is 0?
I assume it won't, but if it somehow can than this will be UB, because rust's ZST's has non-null pointers with zero length

**Comment by ariard** [comment]
@elichai you're thinking about a non-null pointers to a zero-length `spending` arg ? If so `verify_with_flags ` shouldn't be called as there is no inputs to iter on ?

File: src/blockdata/script.rs [hunk 1]
```diff
@@ -346,6 +346,17 @@ impl Script {
...
     }
 
+    #[cfg(feature="bitcoinconsensus")]
+    /// verify spend of an input script
+    /// # Parameters
+    ///  * index - the input index in spending which is spending this transaction
+    ///  * amount - the amount this script guards
+    ///  * spending - the transaction that attempts to spend the output holding this script
+    ///  * flags - the script interpreter flags with which to enforce verification
+    pub fn verify_with_flags (&self, index: usize, amount: u64, spending: &[u8], flags: u32) -> Result<(), Error> {
```
Comment (stevenroose) [inline src/blockdata/script.rs:356]: Not sure if we should use `Amount` here. It's mostly going to be used in the method from `Transaction`, where the raw `u64` would have to be wrapped if we require `Amount` here..

File: src/blockdata/transaction.rs [hunk 2]
```diff
@@ -452,6 +452,24 @@ impl Transaction {
         Ok(())
     }
 
+    #[cfg(feature="bitcoinconsensus")]
+    /// Verify that this transaction is able to spend its inputs
+    /// with provided script interpreter flags. Flags are applied
+    /// globally for all transaction inputs.
+    /// The lambda spent should not return the same TxOut twice!
```
Comment (stevenroose) [inline src/blockdata/transaction.rs:459]: Hmm I recall the conversations on another proposal to add a `verify` method on `Transaction` that had this requirement. I'm not sure how I feel about this requirement.

I think personally I would prefer a message like "this method doesn't check for calls to the `spent` function with the same key twice, so it is up to the user to check for duplicate inputs".

At least that makes it possible to pass `inputHashMap.get` or some generic closure and do the duplicate check before.

**Comment by apoelstra** [comment]
Subsumed by #330

**Pull request closed without merging**